docs/architecture/overview.md

```diff
@@ -46,6 +49,15 @@ This document specifies the architecture for GCP Ingestion as a whole.
 - Should retry transient Cloud Storage errors indefinitely
     - Should use exponential back-off to determine retry timing
 
+### `EcosystemDecryptor`
+
+- Must have access restricted to a limited set of operators
+- Must load a bundle of private keys from Cloud KMS
```

Comment:
It doesn't seem to be possible to get the private keys that are stored in Cloud KMS. It is also mentioned that decryption/encryption needs to be done within KMS (for example mentioned here: https://cloud.google.com/kms/docs/faq#export_keys)

Reply:
This requires some more investigation, then. From my understanding of KMS, one possibility is that you can ask KMS to hand you a session key derived from the actual private key that it's holding, and then you use that locally held session key to do the actual decrypting of payloads. But it's not clear to me whether we can take advantage of that in the context of our particular use case.

If we truly do need access to the exact private key associated with the public key that the client used to encrypt the original value, then indeed it sounds like KMS is not the right thing to use and we'd instead likely want to store the key in GCS with highly restricted access.

Reply:
A KMS will never provide the private key, that is true.  I see a couple of options:

* Use the KMS directly.  My gut reaction is this gets very expensive very quickly, but I don't have any real numbers to extrapolate to that result.
* Generate in software a public/private keypair for actually decrypting payloads (call it the "payload keypair"), encrypt **that** private key using the KMS and store it.  When the payload keypair needs to be used, load it and decrypt it through the KMS.

Reply:
The [pricing for KMS](https://cloud.google.com/kms/pricing) is $0.03 per 10,000 operations. Looking at the number of daily active users for Firefox Accounts, it would get quite expensive ($300-$400 per day).

Reply:
The second option seems the most cost-effective.  It does assume the bundle is loaded (and therefore decrypted) fairly infrequently.

It could be worthwhile to check with the Autograph team (cc @jvehent ).

Reply:
I wouldn't use a KMS directly. You probably want to process pings much faster than the KMS could provide. The typical use of a KMS is to decrypt a known key or secret at initialization, use that decrypted key or secret from memory, and never write it to disk. This is how autograph loads some of its secrets (when they are not stored in an HSM).

The second option is to use an HSM. HSMs are designed for speed, and can handle many thousands operations per seconds at low latency. They aren't that expensive anymore (in the grand scheme of things), but integrating with them can be somewhat tricky, because you need to deal with all the pkcs11 shenanigans.

Autograph uses an AWS CloudHSM for most of the Firefox signing keys, so we have hands-on with that technology. GCP also has a CloudHSM, but we don't have any experience using it.

Reply:
It looks like GCPs CloudHSM has the same pricing as KMS with $0.03 per 10,000 operations (or $0.15 for other key operations): https://cloud.google.com/hsm#pricing

Reply:
Thanks @jvehent .

I think our best course of action is to follow the "known key" option (my first option/jvehent's first option).

The important point is to *never write the decrypted known private key to disk*.  It's fine to write the _*encrypted*_ version of that private key.